enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1012 @@
+
+When building the OVN-Kubernetes IPAM module IP pools, it will be needed to
+take into account these persistent IPs as well. Thus, there will be a resource
+to sync when adding a pod requesting an attachment to a network allowing
+persistent IPs.
+
+Using informers to know the `IPAMClaim`s CRs will mitigate this risk, since
+we will not actively synchronously call the Kubernetes API for reading.
+
+### Drawbacks
+
+## Design Details
+
+### KubeVirt related changes
+
+#### Pod templating
+
+KubeVirt would need to template the pod accordingly - i.e. the
+`NetworkSelectionElements` must feature the `ipam-claim-reference` attribute to
+indicate to the IPAM CNI / CNI which `IPAMClaim` they should use to persist the
+IP allocation across reboots / migrations.
+
+Check the following network selection elements annotation from a KubeVirt
+launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "tenantred",
+    "namespace": "default",
+    "interface": "pod16367aacb67",
+    "ipam-claim-reference": "vm-a.tenantred-attachment"
+  }
+]
+```
+
+#### KubeVirt feature gate
+If a feature gate is introduced, KubeVirt will **always** declare its intent of
+having persistent IP allocations for all its VM secondary networks (i.e. multus
+non default networks).
+
+A more efficient alternative would be to only create the `IPAMClaim` objects if
+the `network-attachment-definition` they point to have the `allowPersistentIPs`
+knob enabled.
+
+#### KubeVirt API changes
+
+The persistent IP allocations represented by the `IPAMClaim` objects should be
+reported in the VM status, allowing the user to easily identify which IPs are
+pinned for the VM.
+
+Thus, we propose the adding the following struct to the
+[VirtualMachineStatus](https://github.com/kubevirt/kubevirt/blob/960484df024ae8cce0f189d5f492667c4d57b250/staging/src/kubevirt.io/api/core/v1/types.go#L1486)
+structure:
+
+```go
+// VirtualMachineStatus represents the status returned by the
+// controller to describe how the VirtualMachine is doing
+type VirtualMachineStatus struct {
+	...
+	NetworkInterfacesStatus []NetworkInterfaceStatus `json:"networkInterfacesStatus,omitempty"`
+}
+
+type NetworkInterfaceStatus {
+	InterfaceName string   `json:"interfaceName,omitempty"`
+	IPs           []string `json:"ips,omitempty"`
+}
+```
+
+**NOTE:** this change is **not** required. It does provide helpful information
+to the VM user.
+
+### OVN-Kubernetes related changes
+When the OVN-Kubernetes `ovnkube-control-plane` pod realizes the network
+selection element features `ipam-claim-references`, it will allocate IP
+addresses from the pools, plus update the existing `IPAMClaim` CR with them.
+
+The next sections feature specific information on how to integrate this new CRD
+in the existing OVN-Kubernetes IP pool.
+
+#### Integrate the `IPAMClaim` CRD with OVN-Kubernetes ip pool
```

Comment:
From the OVN perspective we probably need a bit more detail on how to work with this IPAMClaim:

- Can IP assignment to the IPAMClaim happen lazily? So instead of assigning an IP to a claim when the claim is created, could it be assigned when used? This is not being explicitly or clearly stated but I think we should. 
- Can OVN-K use IPAMClaim for its own allocations, unrelated to the inter-working with kubevirt detailed on this spec? The thing is that we need to assume that OVN will need its own CRDs to do IPAM efficiently, so knowing if this IPAMClaim can be framed or should be framed for that purpose as well could be interesting, aiming to reduce the number of CRDs required.
- Can you add more specific detail on how the network selection element is matched to the IPAMClaim. Maybe I missed this somewhere but I could not find an easy way to correlate this with the examples provided.

Reply:
> The network-selection-element features an added attribute named `ipam-claim-reference` that points to the name of the IPAM claim (it must be in the same namespace of the workload).

Maybe then I am confused because the examples don't match? For `ipam-claim-reference` example you are using `vm-a.tenantred-attachment` but then on the IPAMClaim example the name is `vm-a.tenantred`. Would those have to match?

> Hm. Interesting. Yeah, I guess it could be extended to work like that. I could add a paragraph about that. Would the name IPAMClaim hold for that use case ?

Hmm it would be a stretch. But after thinking about it a bit more, let's forget about this possibility. More trouble than its worth.
 
> Hm, I don't think so. If we want to be lazy, I guess we could always choose another alternative, where ovn-k itself creates the IPAMClaim (only when needed). This option would make KubeVirt's unplug feature impossible to implement (well, impossible is too stirct ... but right now, I don't see how it would do it) .

I am not saying lazily creating the IPAMClaim, I am saying lazily assigning IPs to the IPAMClaim

Reply:
> > The network-selection-element features an added attribute named `ipam-claim-reference` that points to the name of the IPAM claim (it must be in the same namespace of the workload).
> 
> Maybe then I am confused because the examples don't match? For `ipam-claim-reference` example you are using `vm-a.tenantred-attachment` but then on the IPAMClaim example the name is `vm-a.tenantred`. Would those have to match?

Oh !!! Yeah, those would have to match. Good catch, I'll correct the examples. 

> 
> > Hm. Interesting. Yeah, I guess it could be extended to work like that. I could add a paragraph about that. Would the name IPAMClaim hold for that use case ?
> 
> Hmm it would be a stretch. But after thinking about it a bit more, let's forget about this possibility. More trouble than its worth.

OK.

> 
> > Hm, I don't think so. If we want to be lazy, I guess we could always choose another alternative, where ovn-k itself creates the IPAMClaim (only when needed). This option would make KubeVirt's unplug feature impossible to implement (well, impossible is too stirct ... but right now, I don't see how it would do it) .
> 
> I am not saying lazily creating the IPAMClaim, I am saying lazily assigning IPs to the IPAMClaim

I see. Right now, I don't see anything preventing lazy IP allocation.